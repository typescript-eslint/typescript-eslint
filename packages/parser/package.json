{
  "name": "@typescript-eslint/parser",
  "version": "2.21.0",
  "description": "An ESLint custom parser which leverages TypeScript ESTree",
  "main": "dist/parser.js",
  "types": "dist/parser.d.ts",
  "files": [
    "dist",
    "README.md",
    "LICENSE"
  ],
  "engines": {
    "node": "^8.10.0 || ^10.13.0 || >=11.10.1"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/typescript-eslint/typescript-eslint.git",
    "directory": "packages/parser"
  },
  "bugs": {
    "url": "https://github.com/typescript-eslint/typescript-eslint/issues"
  },
  "license": "BSD-2-Clause",
  "keywords": [
    "ast",
    "ecmascript",
    "javascript",
    "typescript",
    "parser",
    "syntax",
    "eslint"
  ],
  "scripts": {
    "build": "tsc -b tsconfig.build.json",
    "clean": "tsc -b tsconfig.build.json --clean",
    "format": "prettier --write \"./**/*.{ts,js,json,md}\" --ignore-path ../../.prettierignore",
    "lint": "eslint . --ext .js,.ts --ignore-path='../../.eslintignore'",
    "test": "jest --coverage",
    "typecheck": "tsc -p tsconfig.json --noEmit"
  },
  "peerDependencies": {
    "eslint": "^5.0.0 || ^6.0.0"
  },
  "dependencies": {
    "@types/eslint-visitor-keys": "^1.0.0",
    "@typescript-eslint/experimental-utils": "2.21.0",
    "@typescript-eslint/typescript-estree": "2.21.0",
    "eslint-visitor-keys": "^1.1.0"
  },
  "devDependencies": {
    "@types/glob": "^7.1.1",
    "@typescript-eslint/shared-fixtures": "2.21.0",
    "glob": "*"
  },
  "peerDependenciesMeta": {
    "typescript": {
      "optional": true
    }
  },
  "funding": {
    "type": "opencollective",
    "url": "https://opencollective.com/typescript-eslint"
  }
}
