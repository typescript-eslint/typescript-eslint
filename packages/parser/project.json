{
  "name": "parser",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "type": "library",
  "implicitDependencies": [],
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["packages/parser/**/*.ts"]
      }
    }
  }
}
