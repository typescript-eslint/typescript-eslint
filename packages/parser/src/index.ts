export { parse, parseForESLint, ParserOptions } from './parser';
export { ParserServices } from '@typescript-eslint/typescript-estree';
export { clearCaches } from '@typescript-eslint/typescript-estree';

// note - cannot migrate this to an import statement because it will make TSC copy the package.json to the dist folder
export const version: string = require('../package.json').version;
