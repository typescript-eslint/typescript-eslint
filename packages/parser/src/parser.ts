import { ParserOptions, TSESTree, Lib } from '@typescript-eslint/types';
import {
  parseAndGenerateServices,
  ParserServices,
  TSESTreeOptions,
  visitorKeys,
} from '@typescript-eslint/typescript-estree';
import { analyze, ScopeManager } from '@typescript-eslint/scope-manager';
import debug from 'debug';
import { ScriptTarget } from 'typescript';

const log = debug('typescript-eslint:parser:parser');

interface ParseForESLintResult {
  ast: TSESTree.Program & {
    range?: [number, number];
    tokens?: TSESTree.Token[];
    comments?: TSESTree.Comment[];
  };
  services: ParserServices;
  visitorKeys: typeof visitorKeys;
  scopeManager: ScopeManager;
}

function validateBoolean(
  value: boolean | undefined,
  fallback = false,
): boolean {
  if (typeof value !== 'boolean') {
    return fallback;
  }
  return value;
}

const LIB_FILENAME_REGEX = /lib\.(.+)\.d\.ts/;
function getLib(services: ParserServices): Lib[] {
  const compilerOptions = services.program.getCompilerOptions();
  if (compilerOptions.lib) {
    return compilerOptions.lib
      .map(lib => {
        const match = LIB_FILENAME_REGEX.exec(lib.toLowerCase());
        if (!match) {
          return null;
        }

        return match[1] as Lib;
      })
      .filter(l => l != null) as Lib[];
  }

  const target = compilerOptions.target ?? ScriptTarget.ES5;
  // https://github.com/Microsoft/TypeScript/blob/59ad375234dc2efe38d8ee0ba58414474c1d5169/src/compiler/utilitiesPublic.ts#L13-L32
  switch (target) {
    case ScriptTarget.ESNext:
      return ['esnext.full'];
    case ScriptTarget.ES2020:
      return ['es2020.full'];
    case ScriptTarget.ES2019:
      return ['es2019.full'];
    case ScriptTarget.ES2018:
      return ['es2018.full'];
    case ScriptTarget.ES2017:
      return ['es2017.full'];
    case ScriptTarget.ES2016:
      return ['es2016.full'];
    case ScriptTarget.ES2015:
      return ['es6'];
    default:
      return ['lib'];
  }
}

function parse(
  code: string,
  options?: ParserOptions,
): ParseForESLintResult['ast'] {
  return parseForESLint(code, options).ast;
}

function parseForESLint(
  code: string,
  options?: ParserOptions | null,
): ParseForESLintResult {
  if (!options || typeof options !== 'object') {
    options = {};
  } else {
    options = { ...options };
  }
  // https://eslint.org/docs/user-guide/configuring#specifying-parser-options
  // if sourceType is not provided by default eslint expect that it will be set to "script"
  if (options.sourceType !== 'module' && options.sourceType !== 'script') {
    options.sourceType = 'script';
  }
  if (typeof options.ecmaFeatures !== 'object') {
    options.ecmaFeatures = {};
  }

  const parserOptions: TSESTreeOptions = {};
  Object.assign(parserOptions, options, {
    useJSXTextNode: validateBoolean(options.useJSXTextNode, true),
    jsx: validateBoolean(options.ecmaFeatures.jsx),
  });

  if (typeof options.filePath === 'string') {
    const tsx = options.filePath.endsWith('.tsx');
    if (tsx || options.filePath.endsWith('.ts')) {
      parserOptions.jsx = tsx;
    }
  }

  /**
   * Allow the user to suppress the warning from typescript-estree if they are using an unsupported
   * version of TypeScript
   */
  const warnOnUnsupportedTypeScriptVersion = validateBoolean(
    options.warnOnUnsupportedTypeScriptVersion,
    true,
  );
  if (!warnOnUnsupportedTypeScriptVersion) {
    parserOptions.loggerFn = false;
  }

  const { ast, services } = parseAndGenerateServices(code, parserOptions);
  ast.sourceType = options.sourceType;

  // automatically apply the libs configured for the program
  if (services.hasFullTypeInformation && options.lib == null) {
    options.lib = getLib(services);
    log('Resolved libs from program: %o', options.lib);
  }

  const analyzeOptions = {
    ecmaVersion: options.ecmaVersion,
    globalReturn: options.ecmaFeatures.globalReturn,
    lib: options.lib,
    sourceType: options.sourceType,
  };
  const scopeManager = analyze(ast, analyzeOptions);

  return { ast, services, scopeManager, visitorKeys };
}

export { parse, parseForESLint, ParserOptions };
