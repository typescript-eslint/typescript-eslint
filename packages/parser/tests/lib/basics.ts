import { Linter } from 'eslint';
import fs from 'fs';
import glob from 'glob';
import * as parser from '../../src/parser';
import {
  createScopeSnapshotTestBlock,
  formatSnapshotName,
} from '../tools/test-utils';

const FIXTURES_DIR = './tests/fixtures/basics';
const testFiles = glob.sync(`${FIXTURES_DIR}/**/*.src.js`);

//------------------------------------------------------------------------------
// Tests
//------------------------------------------------------------------------------

describe('basics', () => {
  testFiles.forEach(filename => {
    const code = fs.readFileSync(filename, 'utf8');
    it(
      formatSnapshotName(filename, FIXTURES_DIR),
      createScopeSnapshotTestBlock(code),
    );
  });

  it('https://github.com/eslint/typescript-eslint-parser/issues/476', () => {
    const linter = new Linter();
    const code = `
export const Price: React.SFC<PriceProps> = function Price(props) {}
`;
    const config: Linter.Config = {
      parser: '@typescript-eslint/parser',
      rules: {
        test: 'error',
      },
    };

    linter.defineParser('@typescript-eslint/parser', parser);
    linter.defineRule('test', {
      create(context: any) {
        return {
          TSTypeReference(node: any) {
            const name = context.getSourceCode().getText(node.typeName);
            context.report({
              node,
              message: 'called on {{name}}',
              data: { name },
            });
          },
        };
      },
    });

    const messages = linter.verify(code, config, { filename: 'issue.ts' });

    expect(messages).toStrictEqual([
      {
        column: 21,
        endColumn: 42,
        endLine: 2,
        line: 2,
        message: 'called on React.SFC',
        nodeType: 'TSTypeReference',
        ruleId: 'test',
        severity: 2,
      },
      {
        column: 31,
        endColumn: 41,
        endLine: 2,
        line: 2,
        message: 'called on PriceProps',
        nodeType: 'TSTypeReference',
        ruleId: 'test',
        severity: 2,
      },
    ]);
  });
});
