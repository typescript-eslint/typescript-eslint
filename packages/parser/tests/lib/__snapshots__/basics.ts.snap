// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basics fixtures/delete-expression.src 1`] = `
Object {
  "$id": 2,
  "block": Object {
    "range": Array [
      0,
      16,
    ],
    "type": "Program",
  },
  "childScopes": Array [
    Object {
      "$id": 1,
      "block": Object {
        "range": Array [
          0,
          16,
        ],
        "type": "Program",
      },
      "childScopes": Array [],
      "functionExpressionScope": false,
      "isStrict": true,
      "references": Array [
        Object {
          "$id": 0,
          "from": Object {
            "$ref": 1,
          },
          "identifier": Object {
            "name": "foo",
            "range": Array [
              7,
              10,
            ],
            "type": "Identifier",
          },
          "kind": "r",
          "resolved": null,
          "writeExpr": undefined,
        },
      ],
      "throughReferences": Array [
        Object {
          "$ref": 0,
        },
      ],
      "type": "module",
      "upperScope": Object {
        "$ref": 2,
      },
      "variableMap": Object {},
      "variableScope": Object {
        "$ref": 1,
      },
      "variables": Array [],
    },
  ],
  "functionExpressionScope": false,
  "isStrict": false,
  "references": Array [],
  "throughReferences": Array [
    Object {
      "$ref": 0,
    },
  ],
  "type": "global",
  "upperScope": null,
  "variableMap": Object {},
  "variableScope": Object {
    "$ref": 2,
  },
  "variables": Array [],
}
`;

exports[`basics fixtures/do-while-statements.src 1`] = `
Object {
  "$id": 6,
  "block": Object {
    "range": Array [
      0,
      59,
    ],
    "type": "Program",
  },
  "childScopes": Array [
    Object {
      "$id": 5,
      "block": Object {
        "range": Array [
          0,
          59,
        ],
        "type": "Program",
      },
      "childScopes": Array [
        Object {
          "$id": 4,
          "block": Object {
            "range": Array [
              29,
              43,
            ],
            "type": "BlockStatement",
          },
          "childScopes": Array [],
          "functionExpressionScope": false,
          "isStrict": true,
          "references": Array [
            Object {
              "$id": 3,
              "from": Object {
                "$ref": 4,
              },
              "identifier": Object {
                "name": "i",
                "range": Array [
                  34,
                  35,
                ],
                "type": "Identifier",
              },
              "kind": "rw",
              "resolved": Object {
                "$ref": 0,
              },
              "writeExpr": Object {
                "range": Array [
                  39,
                  40,
                ],
                "type": "Literal",
              },
            },
          ],
          "throughReferences": Array [
            Object {
              "$ref": 3,
            },
          ],
          "type": "block",
          "upperScope": Object {
            "$ref": 5,
          },
          "variableMap": Object {},
          "variableScope": Object {
            "$ref": 5,
          },
          "variables": Array [],
        },
      ],
      "functionExpressionScope": false,
      "isStrict": true,
      "references": Array [
        Object {
          "$id": 1,
          "from": Object {
            "$ref": 5,
          },
          "identifier": Object {
            "name": "i",
            "range": Array [
              19,
              20,
            ],
            "type": "Identifier",
          },
          "kind": "w",
          "resolved": Object {
            "$ref": 0,
          },
          "writeExpr": Object {
            "range": Array [
              23,
              24,
            ],
            "type": "Literal",
          },
        },
        Object {
          "$id": 2,
          "from": Object {
            "$ref": 5,
          },
          "identifier": Object {
            "name": "i",
            "range": Array [
              51,
              52,
            ],
            "type": "Identifier",
          },
          "kind": "r",
          "resolved": Object {
            "$ref": 0,
          },
          "writeExpr": undefined,
        },
      ],
      "throughReferences": Array [],
      "type": "module",
      "upperScope": Object {
        "$ref": 6,
      },
      "variableMap": Object {
        "i": Object {
          "$ref": 0,
        },
      },
      "variableScope": Object {
        "$ref": 5,
      },
      "variables": Array [
        Object {
          "$id": 0,
          "defs": Array [
            Object {
              "name": Object {
                "name": "i",
                "range": Array [
                  19,
                  20,
                ],
                "type": "Identifier",
              },
              "node": Object {
                "range": Array [
                  19,
                  24,
                ],
                "type": "VariableDeclarator",
              },
              "parent": Object {
                "range": Array [
                  15,
                  25,
                ],
                "type": "VariableDeclaration",
              },
              "type": "Variable",
            },
          ],
          "eslintUsed": undefined,
          "identifiers": Array [
            Object {
              "name": "i",
              "range": Array [
                19,
                20,
              ],
              "type": "Identifier",
            },
          ],
          "name": "i",
          "references": Array [
            Object {
              "$ref": 1,
            },
            Object {
              "$ref": 3,
            },
            Object {
              "$ref": 2,
            },
          ],
          "scope": Object {
            "$ref": 5,
          },
        },
      ],
    },
  ],
  "functionExpressionScope": false,
  "isStrict": false,
  "references": Array [],
  "throughReferences": Array [],
  "type": "global",
  "upperScope": null,
  "variableMap": Object {},
  "variableScope": Object {
    "$ref": 6,
  },
  "variables": Array [],
}
`;

exports[`basics fixtures/identifiers-double-underscore.src 1`] = `
Object {
  "$id": 9,
  "block": Object {
    "range": Array [
      0,
      60,
    ],
    "type": "Program",
  },
  "childScopes": Array [
    Object {
      "$id": 8,
      "block": Object {
        "range": Array [
          0,
          60,
        ],
        "type": "Program",
      },
      "childScopes": Array [
        Object {
          "$id": 5,
          "block": Object {
            "range": Array [
              20,
              36,
            ],
            "type": "ClassDeclaration",
          },
          "childScopes": Array [],
          "functionExpressionScope": false,
          "isStrict": true,
          "references": Array [],
          "throughReferences": Array [],
          "type": "class",
          "upperScope": Object {
            "$ref": 8,
          },
          "variableMap": Object {
            "__Foo": Object {
              "$ref": 4,
            },
          },
          "variableScope": Object {
            "$ref": 8,
          },
          "variables": Array [
            Object {
              "$id": 4,
              "defs": Array [
                Object {
                  "name": Object {
                    "name": "__Foo",
                    "range": Array [
                      26,
                      31,
                    ],
                    "type": "Identifier",
                  },
                  "node": Object {
                    "range": Array [
                      20,
                      36,
                    ],
                    "type": "ClassDeclaration",
                  },
                  "parent": undefined,
                  "type": "ClassName",
                },
              ],
              "eslintUsed": undefined,
              "identifiers": Array [
                Object {
                  "name": "__Foo",
                  "range": Array [
                    26,
                    31,
                  ],
                  "type": "Identifier",
                },
              ],
              "name": "__Foo",
              "references": Array [],
              "scope": Object {
                "$ref": 5,
              },
            },
          ],
        },
        Object {
          "$id": 7,
          "block": Object {
            "range": Array [
              38,
              59,
            ],
            "type": "FunctionDeclaration",
          },
          "childScopes": Array [],
          "functionExpressionScope": false,
          "isStrict": true,
          "references": Array [],
          "throughReferences": Array [],
          "type": "function",
          "upperScope": Object {
            "$ref": 8,
          },
          "variableMap": Object {
            "arguments": Object {
              "$ref": 6,
            },
          },
          "variableScope": Object {
            "$ref": 7,
          },
          "variables": Array [
            Object {
              "$id": 6,
              "defs": Array [],
              "eslintUsed": undefined,
              "identifiers": Array [],
              "name": "arguments",
              "references": Array [],
              "scope": Object {
                "$ref": 7,
              },
            },
          ],
        },
      ],
      "functionExpressionScope": false,
      "isStrict": true,
      "references": Array [
        Object {
          "$id": 3,
          "from": Object {
            "$ref": 8,
          },
          "identifier": Object {
            "name": "__test",
            "range": Array [
              4,
              10,
            ],
            "type": "Identifier",
          },
          "kind": "w",
          "resolved": Object {
            "$ref": 0,
          },
          "writeExpr": Object {
            "range": Array [
              13,
              17,
            ],
            "type": "Literal",
          },
        },
      ],
      "throughReferences": Array [],
      "type": "module",
      "upperScope": Object {
        "$ref": 9,
      },
      "variableMap": Object {
        "__Bar": Object {
          "$ref": 2,
        },
        "__Foo": Object {
          "$ref": 1,
        },
        "__test": Object {
          "$ref": 0,
        },
      },
      "variableScope": Object {
        "$ref": 8,
      },
      "variables": Array [
        Object {
          "$id": 0,
          "defs": Array [
            Object {
              "name": Object {
                "name": "__test",
                "range": Array [
                  4,
                  10,
                ],
                "type": "Identifier",
              },
              "node": Object {
                "range": Array [
                  4,
                  17,
                ],
                "type": "VariableDeclarator",
              },
              "parent": Object {
                "range": Array [
                  0,
                  18,
                ],
                "type": "VariableDeclaration",
              },
              "type": "Variable",
            },
          ],
          "eslintUsed": undefined,
          "identifiers": Array [
            Object {
              "name": "__test",
              "range": Array [
                4,
                10,
              ],
              "type": "Identifier",
            },
          ],
          "name": "__test",
          "references": Array [
            Object {
              "$ref": 3,
            },
          ],
          "scope": Object {
            "$ref": 8,
          },
        },
        Object {
          "$id": 1,
          "defs": Array [
            Object {
              "name": Object {
                "name": "__Foo",
                "range": Array [
                  26,
                  31,
                ],
                "type": "Identifier",
              },
              "node": Object {
                "range": Array [
                  20,
                  36,
                ],
                "type": "ClassDeclaration",
              },
              "parent": null,
              "type": "ClassName",
            },
          ],
          "eslintUsed": undefined,
          "identifiers": Array [
            Object {
              "name": "__Foo",
              "range": Array [
                26,
                31,
              ],
              "type": "Identifier",
            },
          ],
          "name": "__Foo",
          "references": Array [],
          "scope": Object {
            "$ref": 8,
          },
        },
        Object {
          "$id": 2,
          "defs": Array [
            Object {
              "name": Object {
                "name": "__Bar",
                "range": Array [
                  47,
                  52,
                ],
                "type": "Identifier",
              },
              "node": Object {
                "range": Array [
                  38,
                  59,
                ],
                "type": "FunctionDeclaration",
              },
              "parent": null,
              "type": "FunctionName",
            },
          ],
          "eslintUsed": undefined,
          "identifiers": Array [
            Object {
              "name": "__Bar",
              "range": Array [
                47,
                52,
              ],
              "type": "Identifier",
            },
          ],
          "name": "__Bar",
          "references": Array [],
          "scope": Object {
            "$ref": 8,
          },
        },
      ],
    },
  ],
  "functionExpressionScope": false,
  "isStrict": false,
  "references": Array [],
  "throughReferences": Array [],
  "type": "global",
  "upperScope": null,
  "variableMap": Object {},
  "variableScope": Object {
    "$ref": 9,
  },
  "variables": Array [],
}
`;

exports[`basics fixtures/instanceof.src 1`] = `
Object {
  "$id": 2,
  "block": Object {
    "range": Array [
      0,
      17,
    ],
    "type": "Program",
  },
  "childScopes": Array [
    Object {
      "$id": 1,
      "block": Object {
        "range": Array [
          0,
          17,
        ],
        "type": "Program",
      },
      "childScopes": Array [],
      "functionExpressionScope": false,
      "isStrict": true,
      "references": Array [
        Object {
          "$id": 0,
          "from": Object {
            "$ref": 1,
          },
          "identifier": Object {
            "name": "Set",
            "range": Array [
              14,
              17,
            ],
            "type": "Identifier",
          },
          "kind": "r",
          "resolved": null,
          "writeExpr": undefined,
        },
      ],
      "throughReferences": Array [
        Object {
          "$ref": 0,
        },
      ],
      "type": "module",
      "upperScope": Object {
        "$ref": 2,
      },
      "variableMap": Object {},
      "variableScope": Object {
        "$ref": 1,
      },
      "variables": Array [],
    },
  ],
  "functionExpressionScope": false,
  "isStrict": false,
  "references": Array [],
  "throughReferences": Array [
    Object {
      "$ref": 0,
    },
  ],
  "type": "global",
  "upperScope": null,
  "variableMap": Object {},
  "variableScope": Object {
    "$ref": 2,
  },
  "variables": Array [],
}
`;

exports[`basics fixtures/new-with-member-expression.src 1`] = `
Object {
  "$id": 2,
  "block": Object {
    "range": Array [
      0,
      15,
    ],
    "type": "Program",
  },
  "childScopes": Array [
    Object {
      "$id": 1,
      "block": Object {
        "range": Array [
          0,
          15,
        ],
        "type": "Program",
      },
      "childScopes": Array [],
      "functionExpressionScope": false,
      "isStrict": true,
      "references": Array [
        Object {
          "$id": 0,
          "from": Object {
            "$ref": 1,
          },
          "identifier": Object {
            "name": "foo",
            "range": Array [
              4,
              7,
            ],
            "type": "Identifier",
          },
          "kind": "r",
          "resolved": null,
          "writeExpr": undefined,
        },
      ],
      "throughReferences": Array [
        Object {
          "$ref": 0,
        },
      ],
      "type": "module",
      "upperScope": Object {
        "$ref": 2,
      },
      "variableMap": Object {},
      "variableScope": Object {
        "$ref": 1,
      },
      "variables": Array [],
    },
  ],
  "functionExpressionScope": false,
  "isStrict": false,
  "references": Array [],
  "throughReferences": Array [
    Object {
      "$ref": 0,
    },
  ],
  "type": "global",
  "upperScope": null,
  "variableMap": Object {},
  "variableScope": Object {
    "$ref": 2,
  },
  "variables": Array [],
}
`;

exports[`basics fixtures/new-without-parens.src 1`] = `
Object {
  "$id": 5,
  "block": Object {
    "range": Array [
      0,
      23,
    ],
    "type": "Program",
  },
  "childScopes": Array [
    Object {
      "$id": 4,
      "block": Object {
        "range": Array [
          0,
          23,
        ],
        "type": "Program",
      },
      "childScopes": Array [
        Object {
          "$id": 3,
          "block": Object {
            "range": Array [
              0,
              16,
            ],
            "type": "FunctionDeclaration",
          },
          "childScopes": Array [],
          "functionExpressionScope": false,
          "isStrict": true,
          "references": Array [],
          "throughReferences": Array [],
          "type": "function",
          "upperScope": Object {
            "$ref": 4,
          },
          "variableMap": Object {
            "arguments": Object {
              "$ref": 2,
            },
          },
          "variableScope": Object {
            "$ref": 3,
          },
          "variables": Array [
            Object {
              "$id": 2,
              "defs": Array [],
              "eslintUsed": undefined,
              "identifiers": Array [],
              "name": "arguments",
              "references": Array [],
              "scope": Object {
                "$ref": 3,
              },
            },
          ],
        },
      ],
      "functionExpressionScope": false,
      "isStrict": true,
      "references": Array [
        Object {
          "$id": 1,
          "from": Object {
            "$ref": 4,
          },
          "identifier": Object {
            "name": "X",
            "range": Array [
              21,
              22,
            ],
            "type": "Identifier",
          },
          "kind": "r",
          "resolved": Object {
            "$ref": 0,
          },
          "writeExpr": undefined,
        },
      ],
      "throughReferences": Array [],
      "type": "module",
      "upperScope": Object {
        "$ref": 5,
      },
      "variableMap": Object {
        "X": Object {
          "$ref": 0,
        },
      },
      "variableScope": Object {
        "$ref": 4,
      },
      "variables": Array [
        Object {
          "$id": 0,
          "defs": Array [
            Object {
              "name": Object {
                "name": "X",
                "range": Array [
                  9,
                  10,
                ],
                "type": "Identifier",
              },
              "node": Object {
                "range": Array [
                  0,
                  16,
                ],
                "type": "FunctionDeclaration",
              },
              "parent": null,
              "type": "FunctionName",
            },
          ],
          "eslintUsed": undefined,
          "identifiers": Array [
            Object {
              "name": "X",
              "range": Array [
                9,
                10,
              ],
              "type": "Identifier",
            },
          ],
          "name": "X",
          "references": Array [
            Object {
              "$ref": 1,
            },
          ],
          "scope": Object {
            "$ref": 4,
          },
        },
      ],
    },
  ],
  "functionExpressionScope": false,
  "isStrict": false,
  "references": Array [],
  "throughReferences": Array [],
  "type": "global",
  "upperScope": null,
  "variableMap": Object {},
  "variableScope": Object {
    "$ref": 5,
  },
  "variables": Array [],
}
`;

exports[`basics fixtures/typeof-expression.src 1`] = `
Object {
  "$id": 1,
  "block": Object {
    "range": Array [
      0,
      13,
    ],
    "type": "Program",
  },
  "childScopes": Array [
    Object {
      "$id": 0,
      "block": Object {
        "range": Array [
          0,
          13,
        ],
        "type": "Program",
      },
      "childScopes": Array [],
      "functionExpressionScope": false,
      "isStrict": true,
      "references": Array [],
      "throughReferences": Array [],
      "type": "module",
      "upperScope": Object {
        "$ref": 1,
      },
      "variableMap": Object {},
      "variableScope": Object {
        "$ref": 0,
      },
      "variables": Array [],
    },
  ],
  "functionExpressionScope": false,
  "isStrict": false,
  "references": Array [],
  "throughReferences": Array [],
  "type": "global",
  "upperScope": null,
  "variableMap": Object {},
  "variableScope": Object {
    "$ref": 1,
  },
  "variables": Array [],
}
`;

exports[`basics fixtures/update-expression.src 1`] = `
Object {
  "$id": 8,
  "block": Object {
    "range": Array [
      0,
      39,
    ],
    "type": "Program",
  },
  "childScopes": Array [
    Object {
      "$id": 7,
      "block": Object {
        "range": Array [
          0,
          39,
        ],
        "type": "Program",
      },
      "childScopes": Array [
        Object {
          "$id": 6,
          "block": Object {
            "range": Array [
              11,
              34,
            ],
            "type": "FunctionDeclaration",
          },
          "childScopes": Array [],
          "functionExpressionScope": false,
          "isStrict": true,
          "references": Array [
            Object {
              "$id": 5,
              "from": Object {
                "$ref": 6,
              },
              "identifier": Object {
                "name": "i",
                "range": Array [
                  28,
                  29,
                ],
                "type": "Identifier",
              },
              "kind": "rw",
              "resolved": Object {
                "$ref": 0,
              },
              "writeExpr": null,
            },
          ],
          "throughReferences": Array [
            Object {
              "$ref": 5,
            },
          ],
          "type": "function",
          "upperScope": Object {
            "$ref": 7,
          },
          "variableMap": Object {
            "arguments": Object {
              "$ref": 4,
            },
          },
          "variableScope": Object {
            "$ref": 6,
          },
          "variables": Array [
            Object {
              "$id": 4,
              "defs": Array [],
              "eslintUsed": undefined,
              "identifiers": Array [],
              "name": "arguments",
              "references": Array [],
              "scope": Object {
                "$ref": 6,
              },
            },
          ],
        },
      ],
      "functionExpressionScope": false,
      "isStrict": true,
      "references": Array [
        Object {
          "$id": 2,
          "from": Object {
            "$ref": 7,
          },
          "identifier": Object {
            "name": "i",
            "range": Array [
              4,
              5,
            ],
            "type": "Identifier",
          },
          "kind": "w",
          "resolved": Object {
            "$ref": 0,
          },
          "writeExpr": Object {
            "range": Array [
              8,
              9,
            ],
            "type": "Literal",
          },
        },
        Object {
          "$id": 3,
          "from": Object {
            "$ref": 7,
          },
          "identifier": Object {
            "name": "f",
            "range": Array [
              35,
              36,
            ],
            "type": "Identifier",
          },
          "kind": "r",
          "resolved": Object {
            "$ref": 1,
          },
          "writeExpr": undefined,
        },
      ],
      "throughReferences": Array [],
      "type": "module",
      "upperScope": Object {
        "$ref": 8,
      },
      "variableMap": Object {
        "f": Object {
          "$ref": 1,
        },
        "i": Object {
          "$ref": 0,
        },
      },
      "variableScope": Object {
        "$ref": 7,
      },
      "variables": Array [
        Object {
          "$id": 0,
          "defs": Array [
            Object {
              "name": Object {
                "name": "i",
                "range": Array [
                  4,
                  5,
                ],
                "type": "Identifier",
              },
              "node": Object {
                "range": Array [
                  4,
                  9,
                ],
                "type": "VariableDeclarator",
              },
              "parent": Object {
                "range": Array [
                  0,
                  10,
                ],
                "type": "VariableDeclaration",
              },
              "type": "Variable",
            },
          ],
          "eslintUsed": undefined,
          "identifiers": Array [
            Object {
              "name": "i",
              "range": Array [
                4,
                5,
              ],
              "type": "Identifier",
            },
          ],
          "name": "i",
          "references": Array [
            Object {
              "$ref": 2,
            },
            Object {
              "$ref": 5,
            },
          ],
          "scope": Object {
            "$ref": 7,
          },
        },
        Object {
          "$id": 1,
          "defs": Array [
            Object {
              "name": Object {
                "name": "f",
                "range": Array [
                  20,
                  21,
                ],
                "type": "Identifier",
              },
              "node": Object {
                "range": Array [
                  11,
                  34,
                ],
                "type": "FunctionDeclaration",
              },
              "parent": null,
              "type": "FunctionName",
            },
          ],
          "eslintUsed": undefined,
          "identifiers": Array [
            Object {
              "name": "f",
              "range": Array [
                20,
                21,
              ],
              "type": "Identifier",
            },
          ],
          "name": "f",
          "references": Array [
            Object {
              "$ref": 3,
            },
          ],
          "scope": Object {
            "$ref": 7,
          },
        },
      ],
    },
  ],
  "functionExpressionScope": false,
  "isStrict": false,
  "references": Array [],
  "throughReferences": Array [],
  "type": "global",
  "upperScope": null,
  "variableMap": Object {},
  "variableScope": Object {
    "$ref": 8,
  },
  "variables": Array [],
}
`;

exports[`basics fixtures/void-expression.src 1`] = `
Object {
  "$id": 1,
  "block": Object {
    "range": Array [
      0,
      17,
    ],
    "type": "Program",
  },
  "childScopes": Array [
    Object {
      "$id": 0,
      "block": Object {
        "range": Array [
          0,
          17,
        ],
        "type": "Program",
      },
      "childScopes": Array [],
      "functionExpressionScope": false,
      "isStrict": true,
      "references": Array [],
      "throughReferences": Array [],
      "type": "module",
      "upperScope": Object {
        "$ref": 1,
      },
      "variableMap": Object {},
      "variableScope": Object {
        "$ref": 0,
      },
      "variables": Array [],
    },
  ],
  "functionExpressionScope": false,
  "isStrict": false,
  "references": Array [],
  "throughReferences": Array [],
  "type": "global",
  "upperScope": null,
  "variableMap": Object {},
  "variableScope": Object {
    "$ref": 1,
  },
  "variables": Array [],
}
`;
